src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp

```diff
@@ -3218,20 +3182,27 @@ void NativeInvokerGenerator::generate() {
   __ block_comment("{ L_safepoint_poll_slow_path");
   __ bind(L_safepoint_poll_slow_path);
 
-  if (need_spills) {
-    spill_register(ret_reg);
+  // Need to save the native result registers around any runtime calls.
+  RegSet spills, fp_spills;
+  for (int i = 0; i < _output_registers.length(); i++) {
+    VMReg output = _output_registers.at(i);
+    if (output->is_Register()) {
+      spills += RegSet::of(output->as_Register());
+    } else if (output->is_FloatRegister()) {
+      fp_spills += RegSet::of((Register)output->as_FloatRegister());
```

Comment:
This looks very strange. Does it generate the correct code for FloatRegisters?

Reply:
Er... no. But not because of the cast. The `push(fp_spills)` below should be `push_fp(fp_spills)`. I'll add a FloatRegister constructor to RegSet so it doesn't need that any more. There's one other place that does it in cpu/aarch64/gc/z/zBarrierSetAssembler_aarch64.cpp

Reply:
push_fp() doesn't make much sense if the RegSet is a set of Registers, which are by definition not FloatRegisters. That casting of Register to FloatRegister in gc/z is evil.

Reply:
Should we have a separate RegSet type for FloatRegisters to avoid mixing them up?

Reply:
Absolutely. I'd make an AbstractRegSet and use it as a base type
for both RegSet and FloatRegSet, then we can get rid of the casts
altogether.

Reply:
OK I've done that in the latest commit.